apps/prairielearn/src/pages/instructorAssessmentAccessOverrides/instructorAssessmentAccessOverrides.js

```diff
@@ -0,0 +1,132 @@
+// @ts-check
+import * as express from 'express';
+const asyncHandler = require('express-async-handler');
+import * as sqldb from '@prairielearn/postgres';
+import * as error from '@prairielearn/error';
+import { getEnrollmentForUserInCourseInstance } from '../../models/enrollment';
+import { selectUserByUid } from '../../models/user';
+
+const router = express.Router();
+const sql = sqldb.loadSqlEquiv(__filename);
+
+router.get(
+  '/',
+  asyncHandler(async (req, res) => {
+    const result = await sqldb.queryAsync(sql.select_assessment_access_policies, {
+      assessment_id: res.locals.assessment.id,
+      student_uid: req.body.student_uid,
+    });
+    res.locals.policies = result.rows;
+    // console.log(res.locals);
+    res.render(__filename.replace(/\.js$/, '.ejs'), res.locals);
+  }),
+);
+
+router.post(
+  '/',
+  asyncHandler(async (req, res) => {
+    if (req.body.__action === 'add_new_override') {
+      const params = {
+        assessment_id: res.locals.assessment.id,
+        created_by: res.locals.authn_user.user_id,
+        credit: req.body.credit,
+        end_date: new Date(req.body.end_date),
+        group_name: req.body.group_name || null,
+        note: req.body.note || null,
+        start_date: new Date(req.body.start_date),
+        user_id: req.body.student_uid,
+        student_uid: req.body.student_uid,
+        group_id: null,
+      };
+      // First, validate if group belongs to the assessment
+      if (params.group_name) {
+        const group_result = await sqldb.queryZeroOrOneRowAsync(sql.select_group_in_assessment, {
+          group_name: params.group_name,
+          course_instance_id: res.locals.course_instance.id,
+          assessment_id: res.locals.assessment.id,
+        });
+        // Get the group_id from the result
+        if (group_result.rows.length > 0) {
+          params.group_id = group_result.rows[0].id;
+          console.log('GROUP ID: ' + params.group_id);
+        } else {
+          params.group_id = null;
+        }
+
+        // If group does not belong to assessments and indirectly course instances, return error
+        if (!params.group_id) {
+          throw error.make(400, 'Group not found in this assessment.');
+        }
+      } else if (params.user_id) {
+        const user = await selectUserByUid(params.user_id);
+        if (!user) {
+          throw error.make(400, 'Student UID does not exist.');
+        }
```

Comment:
As written, this would allow someone to enumerate which students exist in PrairieLearn, since it differentiates between a user not existing and a user not being enrolled in the course instance. This is also duplicated between the create and edit branches. Consider pulling this out into another function where we can more easily handle the various cases here.

```ts
function userIsEnrolledInCourseInstance({ uid, course_instance_id }) {
  const user = await selectUserByUid(uid);
  if (!user) return false;

  const enrollment = await getEnrollmentForUserInCourseInstance({
    user_id: user.user_id,
    course_instance_id: course_instance_id,
  });
  return !!enrollment;
}

// Later, here in the handler:
const isEnrolled = await userIsEnrolledInCourseInstance({
  uid: req.body.student_uid,
  course_instance_id: res.locals.course_instance.id,
});
if (!isEnrolled) {
  throw error.make(400, `User ${req.body.student_uid} is not enrolled in this course instance.`);
}
```

Reply:
pulled out in 6d6e070278f2e22de9aacb64f3e9d975bcc47983

Reply:
I see a comment later on that is going to require me to get a validated user_id in insert from js and will change this commit